cloudify_agent/api/plugins/installer.py

```diff
@@ -135,67 +142,107 @@ def _install_managed_plugin(self,
             else:
+                raise exceptions.PluginInstallationError(
+                    'Managed plugin installation found but it is '
+                    'in a corrupted state. [{0}]'.format(managed_plugin))
+
+        fields = ['package_name',
+                  'package_version',
+                  'supported_platform',
+                  'distribution',
+                  'distribution_release']
+        description = ', '.join('{0}: {1}'.format(
+            field, managed_plugin.get(field))
+            for field in fields if managed_plugin.get(field))
+
+        if matching_existing_installation:
+            self.logger.info(
+                'Using existing installation of managed plugin: {0} [{1}]'
+                .format(managed_plugin.id, description))
+        elif (deployment_id != SYSTEM_DEPLOYMENT and
+              plugin['executor'] == 'central_deployment_agent'):
+            raise exceptions.PluginInstallationError(
+                'Central deployment agent managed plugins can only be '
+                'installed using the REST plugins API. [{0}]'
+                .format(managed_plugin))
+        else:
+            try:
                 self.logger.info('Installing managed plugin: {0} [{1}]'
                                  .format(managed_plugin.id, description))
-                try:
-                    self._wagon_install(plugin=managed_plugin, args=args)
-                    shutil.move(tmp_plugin_dir, dst_dir)
-                    with open(os.path.join(dst_dir, 'plugin.id'), 'w') as f:
-                        f.write(managed_plugin.id)
-                except Exception as e:
-                    tpe, value, tb = sys.exc_info()
-                    raise NonRecoverableError('Failed installing managed '
-                                              'plugin: {0} [{1}][{2}]'
-                                              .format(managed_plugin.id,
-                                                      plugin, e)), None, tb
-        finally:
-            if lock:
-                lock.release()
+                self._wagon_install(plugin=managed_plugin, args=args)
+                shutil.move(tmp_plugin_dir, dst_dir)
+                with open(os.path.join(dst_dir, 'plugin.id'), 'w') as f:
+                    f.write(managed_plugin.id)
+                # Wait for Syncthing to sync plugin files on all managers
+                if syncthing_utils:
+                    self._wait_for_syncthing()
+                self._update_plugin_status(managed_plugin.id)
+            except Exception as e:
+                tpe, value, tb = sys.exc_info()
+                raise NonRecoverableError('Failed installing managed '
+                                          'plugin: {0} [{1}][{2}]'
+                                          .format(managed_plugin.id,
+                                                  plugin, e)), None, tb
+
+    @staticmethod
+    def _wait_for_syncthing():
```

Comment:
this whole function has to do with waiting for replication so i'd put it in premium

and then this module wouldn't need to know anything about the replication impl (now it needs to know the config path and that it needs to poll)